src/core/common/array_list.hpp

```diff
@@ -0,0 +1,153 @@
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes the definition for array list class template.
+ */
+
+#ifndef ARRAY_LIST_HPP_
+#define ARRAY_LIST_HPP_
+
+#include <stddef.h>
+
+#include <openthread/error.h>
+
+#include "common/code_utils.hpp"
+
+namespace ot {
+template <typename Type> struct EqualTo
+{
+    bool operator()(const Type &lhs, const Type &rhs) const { return lhs == rhs; }
+};
+
+/**
+ * This class implements a list using array.
+ *
+ * @tparam Element   The type to element stored in the list.
+ * @tparam kMaxSize  The max size of the list.
+ *
+ */
+template <typename Element, ptrdiff_t kMaxSize> class ArrayList
+{
+public:
+    using SizeType = ptrdiff_t;
+
+    Element *begin() { return mElements; }
+
+    Element *end() { return mElements + mSize; }
+
+    /**
+     * This method appends the given element to the back of the list.
+     *
+     * @param[in]  aElement The element to be added.
+     *
+     * @retval OT_ERROR_NONE     The element was added successfully.
+     * @retval OT_ERROR_NO_BUFS  The list is full.
+     *
+     */
+    otError Append(const Element &aElement)
+    {
+        otError error = OT_ERROR_NONE;
+
+        if (mSize >= kMaxSize)
+        {
+            ExitNow(error = OT_ERROR_NO_BUFS);
+        }
+
+        mElements[mSize] = aElement;
+        ++mSize;
+
+    exit:
+        return error;
+    }
+
+    /**
+     * This method removes the element at the given index.
+     *
+     * @param[in] aIndex  The index to be removed.
+     *
+     */
+    void Remove(SizeType aIndex)
```

Comment:
Define `Pop()` to remove the last element in the array?

Then we can remove the ith element efficiently using array[i] = array.Pop()